packages/optimistic-oracle/src/proposer.js

```diff
@@ -286,159 +285,152 @@ class OptimisticOracleProposer {
       this.toBN(proposalPrice.toString())
     );
     if (isPriceDisputable) {
-      // Create the transaction.
+      // Get successful transaction receipt and return value or error.
       const dispute = this.optimisticOracleContract.methods.disputePrice(
         priceRequest.requester,
         this.utf8ToHex(priceRequest.identifier),
         priceRequest.timestamp,
         priceRequest.ancillaryData
       );
-
-      // Simple version of inventory management: simulate the transaction and assume that if it fails,
-      // the caller didn't have enough collateral.
-      let disputeBond, gasEstimation;
-      try {
-        [disputeBond, gasEstimation] = await Promise.all([
-          dispute.call({ from: this.account }),
-          dispute.estimateGas({ from: this.account })
-        ]);
-      } catch (error) {
-        this.logger.error({
-          at: "OptimisticOracleProposer#sendDisputes",
-          message: "Cannot dispute price: not enough collateral (or large enough approval)✋",
-          proposer: this.account,
-          disputeBond,
-          priceRequest,
-          error
-        });
-        return;
-      }
       const txnConfig = {
-        from: this.account,
-        gas: Math.min(Math.floor(gasEstimation * this.GAS_LIMIT_BUFFER), this.txnGasLimit),
         gasPrice: this.gasEstimator.getCurrentFastPrice()
       };
-
       this.logger.debug({
         at: "OptimisticOracleProposer#sendDisputes",
         message: "Disputing proposal",
         priceRequest,
         proposalPrice,
-        disputeBond,
         disputePrice,
         allowedError: this.disputePriceErrorPercent,
         txnConfig
       });
+      const transactionResult = await runTransaction(
+        dispute,
+        this.account,
+        txnConfig,
+        this.GAS_LIMIT_BUFFER,
+        this.txnGasLimit
+      );
 
-      // Send the transaction or report failure.
-      let receipt;
-      try {
-        receipt = await dispute.send(txnConfig);
-      } catch (error) {
-        this.logger.error({
+      // Handle error:
+      if (transactionResult.error) {
+        if (transactionResult.error.type === "call") {
```

Comment:
nit: here and elsewhere, I think this should be `transactionResult.type` rather than `transactionResult.error.type` since the type is returned _next to_ the error, not as a field inside of it.

Reply:
Ah yes agreed